### PR TITLE
privval: Query validator key

### PR DESCRIPTION
**Pull request by marbar3778**
## Description

- Query validator key when a remote signer is used. This is supported gRPC remote signing and filePV only. 

Closes: #3009

### CODE REVIEW

File: cmd/tendermint/commands/show_validator.go [hunk 1]
```diff
@@ -5,31 +5,55 @@ import (
...
+	)
 
-	pv := privval.LoadFilePV(keyFilePath, config.PrivValidatorStateFile())
+	//todo: remove once gRPC is the only supported protocol
+	protocol, _ := tmnet.ProtocolAndAddress(config.PrivValidatorListenAddr)
+	switch protocol {
+	case "grpc":
+		pvsc, err := tmgrpc.DialRemoteSigner(config, args[0], logger)
+		if err != nil {
+			return fmt.Errorf("can't instansiate connection: %w", err)
```
Comment (cmwaters) [inline cmd/tendermint/commands/show_validator.go:38]: nit: *instantiate

File: cmd/tendermint/commands/show_validator.go [hunk 2]
```diff
@@ -5,31 +5,55 @@ import (
...
+	tmgrpc "github.com/tendermint/tendermint/privval/grpc"
 )
 
 // ShowValidatorCmd adds capabilities for showing the validator info.
 var ShowValidatorCmd = &cobra.Command{
-	Use:     "show-validator",
+	Use:     "show-validator [chainID]",
 	Aliases: []string{"show_validator"},
 	Short:   "Show this node's validator info",
+	Args:    cobra.ExactArgs(1),
```
Comment (cmwaters) [inline cmd/tendermint/commands/show_validator.go:21]: I believe this means that the user has to add a chainID even if it's not needed (in the case of a "non grpc" privval). I don't know if we want to consider just making the argument optional or using a flag instead (I thought that config already had chainID within it).

Comment (cmwaters) [reply cmd/tendermint/commands/show_validator.go:21]: I think this makes it a breaking change to the CLI

**Review by cmwaters — COMMENTED** [review]
LGTM :+1: 

I have a potentially naive question: The command starts a new connection right? There's no problem with opening as many connections as possible to the remote signer and their being no conflicts in terms of networking ( i.e.: `bind: address already in use`) nor failure for the remote signer to recognise that all these connections belong to the same trusted tendermint process?

File: cmd/tendermint/commands/show_validator.go (hunk 2 above)
Comment (marbar3778) [reply cmd/tendermint/commands/show_validator.go]: Removed the need for the arg by loading it from genesis. No longer breaking 😄

File: cmd/tendermint/commands/show_validator.go [hunk 3]
```diff
@@ -5,31 +5,55 @@ import (
...
-	if !tmos.FileExists(keyFilePath) {
-		return fmt.Errorf("private validator file %s does not exist", keyFilePath)
-	}
+	var (
+		pubKey crypto.PubKey
+		err    error
+	)
 
-	pv := privval.LoadFilePV(keyFilePath, config.PrivValidatorStateFile())
+	//todo: remove once gRPC is the only supported protocol
```
[the remark this answers is not in the document]
Comment (marbar3778) [reply cmd/tendermint/commands/show_validator.go]: https://github.com/tendermint/tendermint/issues/5883

File: cmd/tendermint/commands/show_validator.go [hunk 4]
```diff
@@ -5,31 +5,63 @@ import (
...
+	case "grpc":
+		genFile := config.GenesisFile()
+		if !tmos.FileExists(genFile) {
+			return fmt.Errorf("unable to find genesis file")
+		}
+		genDoc, err := types.LoadGenesisDoc(genFile)
+		if err != nil {
+			return fmt.Errorf("unable to load genesis file %w", err)
+		}
+		pvsc, err := tmgrpc.DialRemoteSigner(config, genDoc.ChainID, logger)
```
Comment (cmwaters) [inline cmd/tendermint/commands/show_validator.go:44]: can we not use `config.ChainID()`?

File: cmd/tendermint/commands/show_validator.go [hunk 5]
```diff
@@ -5,31 +5,63 @@ import (
...
 	tmos "github.com/tendermint/tendermint/libs/os"
 	"github.com/tendermint/tendermint/privval"
+	tmgrpc "github.com/tendermint/tendermint/privval/grpc"
+	"github.com/tendermint/tendermint/types"
 )
 
 // ShowValidatorCmd adds capabilities for showing the validator info.
 var ShowValidatorCmd = &cobra.Command{
-	Use:     "show-validator",
+	Use:     "show-validator [chainID]",
```
Comment (cmwaters) [inline cmd/tendermint/commands/show_validator.go:19]: I think we can remove this now right?

File: cmd/tendermint/commands/show_validator.go (hunk 4 above)
Comment (marbar3778) [reply cmd/tendermint/commands/show_validator.go]: oh, I didn't know this was available. Changed it to use it, thanks!

**Review by melekes — APPROVED** [review]
👍

**Review by cmwaters — APPROVED** [review]
utACK :+1:  

See my comment above regarding if you can foresee any network problems. Also, are you able to test that this works?

**Comment by marbar3778** [comment]
> See my comment above regarding if you can foresee any network problems. Also, are you able to test that this works?

Sorry read it but forgot to reply. 

This isn't a concern for the new privval since Tendermint acts as the client. The bind address issue is applicable if Tendermint were the server. Since we use two-way TLS the client and server need to be aware of each other, unless the node operator allows others to use the same certificate, the remote signer will deny the connection. If a firewall is being used then the server will be aware that connections are coming from the same location.

**Pull request merged**